src/modules/azure.rs

```diff
@@ -0,0 +1,273 @@
+use ini::Ini;
+use std::fs::File;
+use std::io::{BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use super::{Context, Module, RootModuleConfig};
+
+type JValue = serde_json::Value;
+
+use crate::configs::azure::AzureConfig;
+use crate::formatter::StringFormatter;
+
+type SubscriptionId = String;
+type SubscriptionName = String;
+
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("azure");
+    let config = AzureConfig::try_load(module.config);
+
+    if config.disabled {
+        return None;
+    };
+
+    let subscription_id = get_azure_subscription_id(context)?;
+    let subscription_name = get_azure_subscription_name(context, &subscription_id)?;
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        formatter
+            .map_meta(|variable, _| match variable {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "subscription" => Some(Ok(subscription_name.to_string())),
+                _ => None,
+            })
+            .parse(None, Some(context))
+    });
+
+    module.set_segments(match parsed {
+        Ok(segments) => segments,
+        Err(error) => {
+            log::warn!("Error in module `azure`:\n{}", error);
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+fn get_config_file_location(context: &Context) -> Option<PathBuf> {
+    context
+        .get_env("AZURE_CONFIG_DIR")
+        .map(PathBuf::from)
+        .or_else(|| {
+            let mut home = context.get_home()?;
+            home.push(".azure");
+            Some(home)
+        })
+}
+
+fn get_azure_subscription_id(context: &Context) -> Option<SubscriptionId> {
+    let mut config_path = get_config_file_location(context)?;
+    config_path.push("clouds.config");
+
+    let config_file = Ini::load_from_file(config_path.as_path()).ok()?;
+    let azure_cloud_section = config_file.section(Some("AzureCloud")).unwrap();
+    let current_subscription_id = azure_cloud_section.get("subscription").unwrap();
+    Some(current_subscription_id.to_string())
```

Comment:
Please don't use `.unwrap()` here. Use `.ok()?` instead, and add a test to ensure the missing section doesn't cause unwanted behaviour.
In general, some logging here and elsewhere would be nice to have.

Reply:
Added logging throughout the module.